Switch from Chalk to the next trait solver

jackh726 opened the pull request:
At this point, opening up for general review, though I don't think this is fully ready to merge.

Some checklist (non-exhaustive) of things to do before/after merge (feel free to edit):

Before:
- [x] Fix the tidy lints (currently, I see that we need to add doc tests to the new files)
- [x] Squash (or just rework) commits
- [x] Review all the changed tests, open an issue to track the "real" regressions
- [x] Go through the `todo!()`s and `FIXME`s and decide which are needed/useful to fix prior to merge
- [x] Decide what (if anything) should get upstreamed in a separate PR(s) first
- [x] Switch `with_db` calls to `ScopedDb::set_db` (or decide on a better API)

After (before stable):
- [ ] Test across different projects, find things that are "blocking" (probably after merge)
- [ ] Figure out where memory opts can be made

After (any time):
- [ ] Transition things to from `chalk_ir` to `rustc_type_ir`
  - Inference table blocks a lot of things (e.g. normalization)
  - Some queries are "easy" like `layout_of_ty`

Mixed:
- [ ] General cleanup of things
  - [ ] Reduce/remove `allow(unused)`

ChayimFriedman2 commented on crates/query-group-macro/src/queries.rs:
```diff
@@ -48,7 +48,8 @@ impl ToTokens for TrackedQuery {
             .into_iter()
-            .chain(self.lru.map(|lru| quote!(lru = #lru)));
+            .chain(self.lru.map(|lru| quote!(lru = #lru)))
+            .chain(Some(quote!(unsafe(non_update_return_type))));
```
That should be changed before merge, right?

jackh726 replied:
Sure, if someone can point me to how to fix it.

jackh726 commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2106 @@
+                    is_box: data.flags.contains(StructFlags::IS_BOX),
+                    is_manually_drop: data.flags.contains(StructFlags::IS_MANUALLY_DROP),
+                    is_unsafe_cell: data.flags.contains(StructFlags::IS_UNSAFE_CELL),
+                    is_variant_list_non_exhaustive: false,
```
I cleaned up `AdtFlags` and this is no longer there.

jackh726 commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2106 @@
+
+            impl $name {
+                pub fn new_from_iter<'db>(interner: DbInterner<'db>, data: impl IntoIterator<Item = $ty>) -> Self {
+                    $name::new_(interner.db(), data.into_iter().collect::<smallvec::SmallVec<[_; 2]>>())
```
Not sure what you mean? It's a `SmallVec` and we're collecting into it, nothing to shrink.

ChayimFriedman2 replied:
It depends on the iterator, if it doesn't have an exact `size_hint()` there may be extra items allocated, and we're interning this so `shrink_to_fit()` could save memory.

jackh726 commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2106 @@
+        paste::paste! {
+            interned_vec_nolifetime_salsa!($name, $ty, nofold);
+
+            impl<'db> rustc_type_ir::TypeFoldable<DbInterner<'db>> for $name {
```
No? We have to make something new to intern.

jackh726 commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2106 @@
+            #[salsa::interned(no_lifetime, constructor = new_, debug)]
+            pub struct $name {
+                #[returns(ref)]
+                inner_: smallvec::SmallVec<[$ty; 2]>,
```
What do you mean by everything?

ChayimFriedman2 replied:
Every type we intern that is a collection (i.e. shouldn't some be `Vec`, if the small optimization is not important for them)?

jackh726 commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2106 @@
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub struct Span(Option<span::Span>);
```
I have made Span just be `()`, but also removed a bunch of Span args and things.

jackh726 commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2106 @@
+        let Some((last_idx, _)) = variant_data.fields().iter().last() else { return None };
+        let field_types = interner.db().field_types_ns(id);
+
+        let last_ty = field_types[last_idx].clone();
```
Done.

jackh726 commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2106 @@
+            SolverDefId::TraitId(id) => id,
+            _ => unreachable!(),
+        };
+        self.db().trait_signature(id).flags.contains(TraitFlags::AUTO)
```
Good catch, done.

jackh726 commented:
Squashed and rebased version here: https://github.com/jackh726/rust-analyzer/commit/5fa8436404468e84630d2c52733b54391729d682

Will switch this PR after a bit of review, just so people can see commits if they need.

Veykril commented on crates/hir-ty/src/consteval.rs:
```diff
@@ -248,17 +248,19 @@ pub(crate) fn const_eval_query(
-    };
-    let c = interpret_mir(db, body, false, trait_env)?.0?;
-    Ok(c)
+    crate::next_solver::tls::with_db(db, || {
```
I didn't follow the conversations on zulip too much but this seems weird. We are inside a query so why do we need to set the tls here manually. We should be able to just use salsa's attached thread local for this purpose https://github.com/salsa-rs/salsa/blob/master/src/attach.rs where required

Veykril commented on crates/hir-ty/src/next_solver/consts.rs:
```diff
@@ -0,0 +1,376 @@
+        // SAFETY: ¯\_(ツ)_/¯
+        crate::next_solver::tls::with_db_out_of_thin_air(|db| {
+            let inner = &self.kind_(db).0;
+            unsafe { std::mem::transmute(inner) }
```
What is this even transmuting? Both source and target should be 'db already?

Veykril replied:
Oh nvm, the database isn't an arg to the function ...

Veykril commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2107 @@
+
+#[derive(Debug, Copy, Clone, Hash, PartialOrd, Ord, PartialEq, Eq)]
+pub struct DbInterner<'db> {
+    pub(crate) db: &'db (),
```
`PhantomData`?

Veykril commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2107 @@
+            LangItem::PanicNullPointerDereference => return None,
+            LangItem::ReceiverTarget => return None,
+            LangItem::UnsafePinned => todo!(),
+            LangItem::AsyncFnOnceOutput => todo!(),
```
This is already covered

Veykril commented on crates/hir-ty/src/next_solver/interner.rs:
```diff
@@ -0,0 +1,2107 @@
+    impl ScopedDb {
+        pub fn set_db(db: &dyn crate::db::HirDatabase) {
+            DB.replace(Some(unsafe {
+                std::mem::transmute::<_, &'static dyn crate::db::HirDatabase>(db)
+            } as *const dyn crate::db::HirDatabase));
+        }
+    }
+
+    impl Drop for ScopedDb {
+        fn drop(&mut self) {
+            DB.replace(None);
+        }
+    }
+
+    pub fn with_db<T>(db: &dyn crate::db::HirDatabase, f: impl FnOnce() -> T) -> T {
+        DB.replace(Some(
+            unsafe { std::mem::transmute::<_, &'static dyn crate::db::HirDatabase>(db) }
+                as *const dyn crate::db::HirDatabase,
+        ));
+        f()
+    }
+
+    pub fn with_db_out_of_thin_air<T>(f: impl FnOnce(&dyn crate::db::HirDatabase) -> T) -> T {
+        DB.with(move |slot| match *slot.borrow() {
+            None => {
+                panic!("Expected DB at {}", std::backtrace::Backtrace::force_capture());
+            }
+            Some(db) => f(unsafe { &*db }),
+        })
+    }
+
+    pub fn with_opt_db_out_of_thin_air<T>(
+        f: impl FnOnce(Option<&dyn crate::db::HirDatabase>) -> T,
+    ) -> T {
+        DB.with(move |slot| match *slot.borrow() {
+            None => f(None),
+            Some(db) => f(Some(unsafe { &*db })),
+        })
+    }
+}
```
So yea this we should be able to replace with salsa's attached API.

Veykril reviewed (commented):
Just a quick glance

ShoyuVanilla commented on crates/hir-ty/src/layout_nextsolver.rs:
```diff
@@ -0,0 +1,347 @@
+    };
+    let dl = &*target;
+    let cx = LayoutCx::new(dl);
+    //let ty = normalize(db, trait_env.clone(), ty);
```
Don't we need type normalization here?

jackh726 replied:
Yes, this needed github.com/rust-lang/rust-analyzer/pull/20329/commits/1f1cedc4d8de2e933b85ad83410dbbbf1a0709af, which I've went ahead and pushed to this branch.

ShoyuVanilla replied:
Oh, I just thought that we already had normalize query but seems that it doesn't work well here. Didn't mean to ask you for extra works 😅

jackh726 replied:
No worries, I had already done that work. The problem with the existing query is it works on chalk types, not rustc_type_ir types (and infer vars are problem for mapping).

ShoyuVanilla commented on crates/hir-ty/src/tests/regression.rs:
```diff
@@ -1236,17 +1239,17 @@ fn test() {
             10..68 '{     ...   } }': ()
             16..66 'for _ ...     }': fn into_iter<()>(()) -> <() as IntoIterator>::IntoIter
-            16..66 'for _ ...     }': IntoIterator::IntoIter<()>
+            16..66 'for _ ...     }': {unknown}
```
I think this test could use a `// FIXME(next-solver)` comment as well, though it's not strictly necessary since a similar test with a regression already includes it.

I've skimmed through all the test changes, and everything else looks fine. There are two other regressions in `crates/hir-ty/src/tests/simple.rs` and `crates/hir-ty/src/tests/traits.rs`, but they seem related to erroneous GATs, so I don't think they're problematic.

jackh726 replied:
I'll add a FIXME, but strictly speaking this is also *wrong* (`(): IntoIterator` does not hold).

jackh726 commented on crates/ide/src/goto_definition.rs:
```diff
@@ -108,7 +108,7 @@ pub(crate) fn goto_definition(
 
             Some(
-                IdentClass::classify_node(sema, &parent)?
+                salsa::attach(sema.db, || IdentClass::classify_node(sema, &parent))?
```
I did think about this and tried to change `with_db` to
```
let snap = &snap;
salsa::attach(snap, || {
    Cancelled::catch(|| f(snap))
})
```
which results in `Cannot change database mid-query.`

Veykril replied:
Ah okay at least part of this is due to syntax highlighting building a new db for comments, so probably not that problematic. A change for later then.

ChayimFriedman2 commented:
> What benefits does this bring ?

A lot!

 - Improved performance now, and even more in the future as the next trait solver is optimized more (e.g. Jack reported analysis-stats took half the time!)
 - Improved accuracy. Most of our false positive errors come from bugs in Chalk. Switching to the new trait solver may finally enable marking type mismatch diagnostics as no longer experimental.
 - General maintenance, as Chalk is no longer maintained.

ShoyuVanilla commented:
I’ve opened https://github.com/rust-lang/rust-analyzer/issues/20422 to track the test regressions. It looks like we only have a few to fix. 😄

davidbarsky commented:
> @Snowiiii
> 
> > What benefits does this bring ?
> 
> A lot!
> 
> * Improved performance now, and even more in the future as the next trait solver is optimized more (e.g. Jack reported analysis-stats took half the time!)
> * Improved accuracy. Most of our false positive errors come from bugs in Chalk. Switching to the new trait solver may finally enable marking type mismatch diagnostics as no longer experimental.
> * General maintenance, as Chalk is no longer maintained.

To expand on "improved performance", the work that analysis-stats _mostly_ measures correspond to things like rust-analyzer's native diagnostics and autocomplete. Features like the initial indexing at startup or code navigation _are not_ impacted by the new trait solver.

ChayimFriedman2 commented:
Indeed it's not startup, but it basically impacts everything else.

Veykril commented:
Alright let's give this a spin!
